File: src/com/aslan/module/cipher/algorithms/DC140713Algorithm_V1.js
import AbstractAlgorithm from "./AbstractAlgorithm";
import {ALGORITHMS} from "../consts";

export default class DC140713Algorithm extends AbstractAlgorithm {

	enc(M, K) {
		let BOX = this.BOX;
		let R = this.R;
		let N = this.N;
		let x = 0, v = R, h = this.H - 1, X = this.H;
		for (let i = 1; i <= R; i++) {
			let U = K[i - 1];
			let V = K[--v];
			for (let y = this.H; y < N;) {
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
				M[y] = BOX[M[y] ^ U[y]] ^ M[x];
				M[x] = BOX[M[x] ^ V[x]] ^ M[y];
				x = (++x) & h, y++;
			}
			x = X >> i;
		}
	}

	dec(C, K) {
		let BOX = this.BOX;
		let R = this.R;
		let N = this.N;
		let x = 0, v = R - 1, h = this.H - 1;
		for (let i = 0; i < R; i++ , v--) {
			let V = K[i];
			let U = K[v];
			for (let y = this.H; y < N;) {
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
				C[x] = BOX[C[x] ^ C[y]] ^ V[x];
				C[y] = BOX[C[x] ^ C[y]] ^ U[y];
				x = (++x) & h, y++;
			}
			x = (1 << i) & h;
		}
	}

	get identity() {
		return ALGORITHMS.DC140713;
	}

	get version() {
		return 1;
	}

}

File: src/com/aslan/module/cipher/algorithms/DC140713Algorithm_V1.java
package com.aslan.module.cipher.algorithms;

import com.aslan.module.cipher.Consts;

public class DC140713Algorithm_V1 extends AbstractAlgorithm {

    public void enc(byte[] M, int o, byte[][] K) {
        int x = 0, v = R, h = this.H - 1;
        for (int i = 1; i <= R; i++) {
            byte[] U = K[i - 1];
            byte[] V = K[--v];
            for (int y = this.H; y < N; ) {
                M[o + y] = (byte) (BOX[M[o + y] ^ U[o + y]] ^ M[o + x]);
                M[o + x] = (byte) (BOX[M[o + x] ^ V[o + x]] ^ M[o + y]);
                x = (++x) & h;
                y++;
            }
            x = H >> i;
        }
    }

    public void dec(byte[] C, int o, byte[][] K) {
        int x = 0, v = R - 1, h = this.H - 1;
        for (int i = 0; i < R; i++, v--) {
            byte[] V = K[i];
            byte[] U = K[v];
            for (int y = this.H; y < N; ) {
                C[o + x] = (byte) (BOX[C[o + x] ^ C[o + y]] ^ V[o + x]);
                C[o + y] = (byte) (BOX[C[o + x] ^ C[o + y]] ^ U[o + y]);
                x = (++x) & h;
                y++;
            }
            x = (1 << i) & h;
        }
    }

    public int identity() {
        return Consts.ALGORITHMS.DC140713;
    }

    public int version() {
        return 1;
    }

}